### PR TITLE
Add initialisation routine for offline drivers

### PR DESCRIPTION
**Pull request by SeanBryan51**
Whilst looking into merging the MPI master and worker drivers (#358), @micaeljtoliveira and I have identified that the model initialisation logic is duplicated across all drivers (serial and MPI master/worker). This change adds a preliminary initialisation routine to remove code duplication and improve code clarity across each of the drivers. The remaining initialisation code in the offline drivers will be moved to this routine in later pull requests to clean up the drivers further.

## Checklist

- [x] The new content is accessible and located in the appropriate section.
- [x] I have checked that links are valid and point to the intended content.
- [x] I have checked my code/text and corrected any misspellings

----
📚 Documentation preview 📚: https://cable--440.org.readthedocs.build/en/440/

### CODE REVIEW

File: src/offline/cable_driver_init.F90
Comment (micaeljtoliveira) [inline src/offline/cable_driver_init.F90:45]: This module could be a good place to stop using module variables for passing data between function, which is considered bad practice. Instead, data should be passed as function arguments.

File: src/offline/cable_driver_init.F90
Comment (micaeljtoliveira) [inline src/offline/cable_driver_init.F90:100]: Maybe a good place to start introducing the `newunit` specifier?

```suggestion
    OPEN(newunit=unit, FILE=CABLE_NAMELIST, STATUS="OLD", ACTION="READ")
    READ(unit, NML=CABLE)
    CLOSE(unit)
```
with `unit` being an integer to be declared above. 

This is safer and easier to read (here, `10` is a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)))

File: src/offline/cable_driver_init.F90
Comment (micaeljtoliveira) [inline src/offline/cable_driver_init.F90:108]: Another good practice that you might consider using here:
```suggestion
      WRITE (error_unit,*) 'This program needs at least 2 processes to run!'
```
This requires adding:
```fortran
USE iso_fortran_env, ONLY : error_unit
```
along the other use statements above.

File: src/offline/cable_mpicommon.F90
Comment (micaeljtoliveira) [inline src/offline/cable_mpicommon.F90:35]: I think you can leave this here for now. Later on we should introduce an object to wrap this kind of MPI information.

**Review by micaeljtoliveira — APPROVED** [review]
@SeanBryan51 Overall looks good!

Most of my comments are quite general and, at this point, they are all optional.

File: src/offline/cable_driver_init.F90
Comment (SeanBryan51) [reply src/offline/cable_driver_init.F90:45]: FYI I initially considered defining these module variables as function arguments to `cable_driver_init` but I realised the variables were not defined at the top level scope of `cable_mpidrv.F90` which prevents me from passing in these variables as arguments. We could have the master and worker drivers call `cable_driver_init` separately instead of in `cable_mpidrv.F90` so that the variables were in scope but we would need to take out the MPI specific initialisation out of `cable_driver_init` as it would be called after the split between master and worker. Given the other namelist variables are module variables, I figured that would be the easiest workaround.

File: src/offline/cable_driver_init.F90
Comment (SeanBryan51) [reply src/offline/cable_driver_init.F90:108]: I will leave this for now as @Whyborn is adding this in as part of https://github.com/CABLE-LSM/CABLE/pull/438

File: src/offline/cable_driver_init.F90
Comment (micaeljtoliveira) [reply src/offline/cable_driver_init.F90:45]: Thanks for the explanation. Let's leave this as it is for now. We can revisit this when all drivers are merged.

File: src/offline/cable_driver_init.F90
Comment (micaeljtoliveira) [reply src/offline/cable_driver_init.F90:100]: Lets continue this discussion in the issue so that it doesn't block this PR.

File: CMakeLists.txt
Comment (micaeljtoliveira) [inline CMakeLists.txt:166]: I don't think this works as intended. Basically there are two rules to generate the same object file and associated module file, one for serial and one for MPI. This means that when they are recompiled the second time, they object file from the first compilation is overwritten.

Compilation when using Makefiles works, but more by accident than by design. On the other hand, Ninja simply refuses to compile the files and rightly so.

I think the correct solution here is to have all the serial and MPI specific objects and module files being generated in different folders.

**Review by micaeljtoliveira — CHANGES_REQUESTED** [review]
I think the issue with the CMake mentioned above needs to be fixed before merging this.

File: CMakeLists.txt
Comment (SeanBryan51) [reply CMakeLists.txt:166]: For reference [this issue](https://gitlab.kitware.com/cmake/cmake/-/issues/22898) looks like it describes what is happening in our case. It provides a workaround using the [Fortran_MODULE_DIRECTORY](https://cmake.org/cmake/help/latest/prop_tgt/Fortran_MODULE_DIRECTORY.html) target property. Micael has proposed an alternative way to structure the CMakeLists.txt so we avoid this issue entirely.

Comment (micaeljtoliveira) [reply CMakeLists.txt:166]: I would suggest to merge this PR as is and I'll then create a new PR with further changes that will have the side-effect of fixing this issues.

Comment (micaeljtoliveira) [reply CMakeLists.txt:166]: New branch with proposed changes here: https://github.com/CABLE-LSM/CABLE/tree/introduce_mpi_group Relevant commit: b9e6056e636fe077ad2ddabd53a5f3f1d3661a2d

This still needs to some to be properly documented.

**Review by micaeljtoliveira — APPROVED** [review]

**Review by ccarouge — APPROVED** [review]

**Comment by SeanBryan51** [comment]
Testing with `benchcab fluxsite -v` shows this branch bitwise reproduces model outputs from the main branch:

PBS log file:
```
2024-11-07 12:29:39,954 - INFO - benchcab.benchcab.py:369 - Running comparison tasks...
2024-11-07 12:29:39,977 - INFO - benchcab.benchcab.py:370 - tasks: 168 (models: 2, sites: 42, science configurations: 4)
2024-11-07 12:32:37,432 - INFO - benchcab.benchcab.py:381 - 0 failed, 168 passed
```
Benchcab version: `4.1.0`

`config.yaml`:
```
realisations:
  - repo:
      git:
        branch: main
  - repo:
      git:
        branch: 425-offline-driver-initialisation-refactor

modules: [
  intel-compiler/2021.1.1,
  netcdf/4.7.4,
  openmpi/4.1.0
]
```

**Pull request merged**
